aiidalab_widgets_base/computational_resources.py

```diff
@@ -375,126 +437,131 @@ def _write_ssh_config(self, private_key_abs_fname=None):
-                        child.sendline(pwd.value)
-
-                    elif index == 1:
-                        print("Success.")
-                        if on_success:
-                            on_success()
-                        break
-
-                    elif index == 2:
-                        print("Keys are already present on the remote machine.")
-                        if on_success:
-                            on_success()
-                        break
-
-                    elif index == 3:  # Adding a new host.
-                        child.sendline("yes")
-
-                    elif index == 4:
-                        print(
-                            "Failed\nLooks like the key pair is not present in ~/.ssh folder."
-                        )
-                        break
-
-                    elif index == 5:
-                        print("Failed\nUnknown hostname.")
-                        break
-
-                    elif index == 6:
-                        print("Failed\nConnection refused.")
-                        break
+        # If hostname & username are not provided - do not do anything.
+        if self.hostname.value == "":  # check hostname
+            self.message = "Please specify the computer hostname."
+            return False
 
-                    else:
-                        print("Failed\nUnknown problem.")
-                        print(child.before, child.after)
-                        break
-                child.close()
-                yield
+        if self.username.value == "":  # check username
+            self.message = "Please specify your SSH username."
+            return False
 
+        private_key_abs_fname = None
+        if self._verification_mode.value == "private_key":
+            # unwrap private key file and setting temporary private_key content
+            private_key_abs_fname, private_key_content = self._private_key
+            if private_key_abs_fname is None:  # check private key file
+                self.message = "Please upload your private key file."
+                return False
+
+            # Write private key in ~/.ssh/ and use the name of upload file,
+            # if exist, generate random string and append to filename then override current name.
+            self._add_private_key(private_key_abs_fname, private_key_content)
+
+        if not self._is_in_config():
+            self._write_ssh_config(private_key_abs_fname=private_key_abs_fname)
+
+        # Copy public key on the remote computer.
+        ssh_connection_thread = threading.Thread(target=self._ssh_copy_id)
+        ssh_connection_thread.start()
+
+    def _ssh_copy_id(self):
+        """Run the ssh-copy-id command and follow it until it is completed."""
+        timeout = 30
+        self.password_message = f"Sending public key to {self.hostname.value}... "
+        self._ssh_connection_process = pexpect.spawn(
+            f"ssh-copy-id {self.hostname.value}"
```

Comment:
I assume it is the case since that's how we are using it here, but does ssh-copy-id automatically copy the public key to the jump hosts as well?

Reply:
This is tested and yes, it will copy key to jump host as well.